### PR TITLE
[SPARK-38676][SQL] Provide SQL query context in runtime error message of Add/Subtract/Multiply

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Provide SQL query context in runtime error of Add/Subtract/Multiply if the data type is Int or Long. This is the first PR for improving the runtime error message. There are more to be done after we decide what the error message should be like.

Before changes:
```
> SELECT i.f1 - 100, i.f1 * smallint('2') AS x FROM INT4_TBL i
java.lang.ArithmeticException
integer overflow. If necessary set spark.sql.ansi.enabled to false (except for ANSI interval type) to bypass this error.
```
After changes:
```
> SELECT i.f1 - 100, i.f1 * smallint('2') AS x FROM INT4_TBL i
java.lang.ArithmeticException
integer overflow. If necessary set spark.sql.ansi.enabled to false (except for ANSI interval type) to bypass this error.

== SQL(line 1, position 25) ==
SELECT '' AS five, i.f1, i.f1 * smallint('2') AS x FROM INT4_TBL i
                         ^^^^^^^^^^^^^^^^^^^^
```

### Why are the changes needed?

Provide SQL query context of runtime errors to users, so that they can understand it better.

### Does this PR introduce _any_ user-facing change?

Yes, improve the runtime error message of Add/Subtract/Multiply

### How was this patch tested?

UT

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/trees/TreeNodeSuite.scala [hunk 1]
```diff
@@ -856,4 +856,36 @@ class TreeNodeSuite extends SparkFunSuite with SQLHelper {
...
+      startPosition = Some(38),
+      startIndex = Some(47),
+      stopIndex = Some(77),
+      sqlText = Some(text),
+      objectType = Some("VIEW"),
+      objectName = Some("some_view"))
+    val expected =
+      """
+        |== SQL of VIEW some_view(line 3, position 38) ==
+        |...7890 + 1234567890 + 1234567890, cast('a'
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/trees/TreeNodeSuite.scala:881]: We can improve this later to also do truncation for the expression SQL string. For example, a better error message in this case is 
```
== SQL of VIEW some_view(line 3, position 38) ==
...7890 + 1234567890 + 1234567890, cast('a'...int), 1234567890 + 1234567890 + 12345...
                                   ^^^^^^^^^^^^^^^
```

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/trees/TreeNodeSuite.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/trees/TreeNodeSuite.scala:881]: Another way is to remove all the "\n" within the current node context:
== SQL of VIEW some_view(line 3, position 38) ==
...7890 + 1234567890 + 1234567890, cast('a' as /* comment */ int), 1234567890 + 1234567890 + 12345...
                                                                  ^^^^^^^^^^^^^^^^^^^^^^

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/trees/TreeNodeSuite.scala:881]: Yea this helps in this case, but the string can be super long and it's still better to do truncation

**Comment by gengliangwang** [comment]
Merging to master/3.3

**Pull request closed without merging**